packages/steps/src/inputFunctions.ts

```diff
@@ -0,0 +1,49 @@
+import crypto from 'crypto';
+import fs from 'fs';
+import path from 'path';
+
+import fg from 'fast-glob';
+
+import { BuildStepGlobalContext } from './BuildStepContext.js';
+
+export async function hashFilesAsync(
+  ctx: BuildStepGlobalContext,
+  files: string[]
+): Promise<string> {
+  const hash = crypto.createHash('sha256');
+
+  await Promise.all(
+    files.map(async (file) => {
+      let filePath = path.join(ctx.defaultWorkingDirectory, file);
+      if (fs.existsSync(filePath) && fs.lstatSync(filePath).isDirectory()) {
+        filePath += '/**/*';
+      }
+      const filePaths = await fg(filePath, { onlyFiles: true });
+      return Promise.all(
+        filePaths.map(async (filePath) => {
+          const content = fs.readFileSync(filePath);
+          ctx.baseLogger.debug(`Hashing file ${filePath}`);
+          hash.update(path.relative(ctx.defaultWorkingDirectory, filePath));
+          return hash.update(content);
+        })
+      );
+    })
+  );
+
+  return hash.digest('hex');
+}
+
+export const hashFiles = hashFilesAsync;
+
+export default async function callInputFunctionAsync(
```

Comment:
This (and changes in `template.ts`) are very cool, good job figuring this out! However:

1. Can we not include it in this pull request already? v0 of this pull request may only allow static strings and cache keys and that's ok.
2. Instead of this manual expression parsing and function calling I think we should try to use `jsepEval` on expressions. [This](https://github.com/expo/eas-build/blob/6376bdc493319be4d43e62d1b751fd2161e73cf4/packages/steps/src/BuildStep.ts#L297-L315) is how we're providing functions to `ifCondition`s. It should be simpler and more consistent with other places this way.